packages/plugin-npm/sources/npmHttpUtils.ts

```diff
@@ -64,15 +66,170 @@ export function getIdentUrl(ident: Ident) {
+      configuration,
+      registry,
+      ident,
+      headers: {
+        ...headers,
+        // We set both headers in case a registry doesn't support ETags
+        [`If-None-Match`]: cached?.etag,
+        [`If-Modified-Since`]: cached?.lastModified,
+      },
+      wrapNetworkRequest: async executor => async () => {
+        const response = await executor();
+
+        if (response.statusCode === 304) {
+          if (cached === null)
+            throw new Error(`Assertion failed: cachedMetadata should not be null`);
+
+          return {
+            ...response,
+            body: cached.metadata,
+          };
+        }
+
+        const packageMetadata = pickPackageMetadata(JSON.parse(response.body.toString()));
+
+        PACKAGE_METADATA_CACHE.set(ident.identHash, packageMetadata);
+
+        const metadata: CachedMetadata = {
+          metadata: packageMetadata,
+          etag: response.headers.etag,
+          lastModified: response.headers[`last-modified`],
+        };
+
+        // We append the PID because it is guaranteed that this code is only run once per process for a given ident
+        const identPathTemp = `${identPath}-${process.pid}.tmp` as PortablePath;
+
+        await xfs.mkdirPromise(registryFolder, {recursive: true});
+        await xfs.writeJsonPromise(identPathTemp, metadata, {compact: true});
+
+        // Doing a rename is important to ensure the cache is atomic
+        await xfs.renamePromise(identPathTemp, identPath);
+
+        return {
+          ...response,
+          body: packageMetadata,
+        };
+      },
+    });
+  });
+}
+
+type CachedMetadata = {
+  metadata: PackageMetadata;
+  etag?: string;
+  lastModified?: string;
+};
+
+export type PackageMetadata = {
+  'dist-tags': Record<string, string>;
+  versions: Record<string, any>;
+};
+
+const CACHED_FIELDS = [
+  `name`,
+
+  `deprecated`,
+  `dist.tarball`,
+
+  `bin`,
+  `scripts`,
+
+  `os`,
+  `cpu`,
+  `libc`,
+
+  `dependencies`,
+  `dependenciesMeta`,
+  `optionalDependencies`,
+
+  `peerDependencies`,
+  `peerDependenciesMeta`,
+];
+
+function pickPackageMetadata(metadata: PackageMetadata): PackageMetadata {
+  return {
+    'dist-tags': metadata[`dist-tags`],
+    versions: Object.fromEntries(Object.entries(metadata.versions).map(([key, value]) => [
+      key,
+      pick(value, CACHED_FIELDS),
+    ])),
+  };
+}
+
+/**
+ * Used to invalidate the on-disk cache when the format changes.
+ */
+const CACHE_KEY = hashUtils.makeHash(...CACHED_FIELDS).slice(0, 6);
+
+function getRegistryFolder(configuration: Configuration, registry: string) {
+  const metadataFolder = getMetadataFolder(configuration);
+
+  const parsed = new URL(registry);
+  const registryFilename = toFilename(parsed.hostname);
+
+  return ppath.join(metadataFolder, CACHE_KEY as Filename, registryFilename);
+}
+
+function getMetadataFolder(configuration: Configuration) {
+  return ppath.join(configuration.get(`globalFolder`), `npm-metadata`);
```

Comment:
I'm thinking perhaps `<global>/cache/metadata` might be a better location, what do you think? Actions that already cache the `<global>/cache` folder would benefit from that out of the box, rather than having to add a new path (or cache the whole `<global>`, which may not be as common).

I'd also suggest moving the `npm-` prefix from the folder to each individual file name, by consistency with the cache itself (which contains files from the npm, git, http fetchers, etc).

Reply:
Ping on this comment?

Reply:
(Sorry for the delay.)

> I'm thinking perhaps `<global>/cache/metadata` might be a better location, what do you think?

I don't really like the idea, the mirror and the npm metadata cache are 2 separate things, and merging them in a single folder would lead to confusion. We've also got `index` which would have to be moved to `cache` too with this line of thinking.

This would just complicate things like `yarn cache clean`, where `--mirror` would have to stop removing the entire folder and just remove the archives.

I'd prefer all of them to be in separate folders. One thing we could do would be to move the mirror to `<global>/cache/mirror` in addition to what you proposed, then I'd be fine with it and it would still benefit from existing caching (but it would look a bit weird to have the global cache in `<global>/cache/mirror`, but I guess that's just the consequence of the mirror and the global cache being the same thing).

> I'd also suggest moving the npm- prefix from the folder to each individual file name, by consistency with the cache itself (which contains files from the npm, git, http fetchers, etc).

I'd rather not. It would just give the illusion of consistency.

The cache can do it because it's the sole source that controls that folder.

The npm metadata cache is supposed to be specific to the `plugin-npm` resolvers - they are the only ones that populate and control it.

Moving the prefix to the filenames and having the folder called just `metadata` gives the illusion that the folder is a generalized metadata cache. And it could be indeed, if other resolvers ever need to cache metadata in it, but it would have different shapes (and possibly different filename formats) based on the resolvers that cache the data and that might lead to confusion.

It would also have to be controlled by the core, which would be tasked with automatically generating the paths to ensure that no 2 resolvers accidentally use the same path (and also to make it possible for us to change the metadata cache path in the future).

In addition, we're not even certain that the current kind of cache is better than a monolithic one, that's something I'm open to experimenting with in the future.

That's why I think that opening the folder to anything but the npm resolvers is not something I want to do yet, if ever.

Reply:
:thinking: Thought more about it and I'd be open to making it `<global>/metadata/npm/<hash>/<registry>/lodash.json`.

This way, we still have a common metadata folder but we make it clear that each resolver has to manage it manually.

What do you think?

Reply:
That seems reasonable. I like the idea of moving the cache into cache/mirror (that said, I think we could also just rename --mirror into -g for the same effect; I think I'd like this even better 🤔).

Reply:
:thinking: I still think I'd prefer to keep the mirror, metadata, and index separate.

> I think we could also just rename --mirror into -g for the same effect

We could have a `-g` too but I prefer to retain the granularity. `yarn cache clean` will become interactive in a future PR anyways (just like we discussed some time ago), and it will clean everything by default.